tsdb/tsm1/engine_measurement_notime_schema.go

```diff
@@ -240,3 +241,112 @@ func (e *Engine) fieldsNoTime(ctx context.Context, orgID, bucketID influxdb.ID,
 
 	return cursors.NewMeasurementFieldsSliceIterator([]cursors.MeasurementFields{{Fields: vals}}), nil
 }
+
+// MeasurementTagKeysNoTime returns an iterator which enumerates the tag keys
+// for the given bucket, measurement and tag key and filtered using the optional
+// the predicate.
+//
+// MeasurementTagKeysNoTime will always return a StringIterator if there is no error.
+//
+// If the context is canceled before MeasurementTagKeysNoTime has finished
+// processing, a non-nil error will be returned along with statistics for the
+// already scanned data.
+func (e *Engine) MeasurementTagKeysNoTime(ctx context.Context, orgID, bucketID influxdb.ID, measurement string, predicate influxql.Expr) (cursors.StringIterator, error) {
+	if measurement != "" {
+		predicate = AddMeasurementToExpr(measurement, predicate)
+	}
+	return e.tagKeysNoTime(ctx, orgID, bucketID, predicate)
+}
+
+func (e *Engine) tagKeysNoTime(ctx context.Context, orgID, bucketID influxdb.ID, predicate influxql.Expr) (cursors.StringIterator, error) {
+	if err := ValidateTagPredicate(predicate); err != nil {
+		return nil, err
+	}
+
+	orgBucket := tsdb.EncodeName(orgID, bucketID)
+
+	vals := make([]string, 0, 32)
+
+	span := opentracing.SpanFromContext(ctx)
+	if span != nil {
+		defer func() {
+			span.LogFields(
+				log.Int("values_count", len(vals)),
+			)
+		}()
+	}
+
+	var (
+		km   keyMerger
+		keys = make([][]byte, 0, 32)
+	)
+
+	if err := func() error {
+		sitr, err := e.index.MeasurementSeriesByExprIterator(orgBucket[:], predicate)
+		if err != nil {
+			return err
+		}
+		defer sitr.Close()
+
+		for i := 0; ; i++ {
+			// to keep cache scans fast, check context every 'cancelCheckInterval' iterations
+			if i%cancelCheckInterval == 0 {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+				}
+			}
+
+			elem, err := sitr.Next()
+			if err != nil {
+				return err
+			} else if elem.SeriesID.IsZero() {
+				return nil
+			}
+
+			sf := e.index.SeriesFile()
+			if sf == nil {
+				return nil
+			}
+
+			skey := sf.SeriesKey(elem.SeriesID)
+			if len(skey) == 0 {
+				continue
+			}
+
+			keys = parseSeriesKeys(skey, keys)
+			km.MergeKeys(keys)
+		}
+		return nil
+	}(); err != nil {
+		return cursors.NewStringSliceIterator(nil), err
+	}
+
+	for _, v := range km.Get() {
+		vals = append(vals, string(v))
+	}
+	sort.Strings(vals)
```

Comment:
This should no longer be necessary, as `km.Get()` should return the keys already sorted